doozerlib/assembly.py

```diff
@@ -0,0 +1,138 @@
+    and the input group config.
+    :param releases_config: A Model for releases.yaml.
+    :param assembly: The name of the assembly
+    :param group_config: The group config to merge into a new group config (original Model will not be altered)
+    """
+    if not assembly or not isinstance(releases_config, Model):
+        return group_config
+
+    target_assembly = releases_config.releases[assembly].assembly
+
+    if target_assembly.basis.assembly:  # Does this assembly inherit from another?
+        # Recursive apply ancestor assemblies
+        group_config = assembly_group_config(releases_config, target_assembly.basis.assembly, group_config)
+
+    target_assembly_group = target_assembly.group
+    if not target_assembly_group:
+        return group_config
+
+    return Model(dict_to_model=merger(target_assembly_group.primitive(), group_config.primitive()))
+
+
+def assembly_metadata_config(releases_config: Model, assembly: str, meta_type: str, distgit_key: str, meta_config: Model) -> Model:
+    """
+    Returns a group config based on the assembly information
+    and the input group config.
+    :param releases_config: A Model for releases.yaml.
+    :param assembly: The name of the assembly
+    :param meta_type: 'rpm' or 'image'
+    :param distgit_key: The name of the underlying component
+    :param meta_config: The meta's config object
+    :return: Returns a computed config for the metadata (e.g. value for meta.config).
+    """
+    if not assembly or not isinstance(releases_config, Model):
+        return meta_config
+
+    target_assembly = releases_config.releases[assembly].assembly
+
+    if target_assembly.basis.assembly:  # Does this assembly inherit from another?
+        # Recursive apply ancestor assemblies
+        meta_config = assembly_metadata_config(releases_config, target_assembly.basis.assembly, meta_type, distgit_key, meta_config)
+
+    config_dict = meta_config.primitive()
+
+    component_list = target_assembly.members[f'{meta_type}s']
+    for component_entry in component_list:
+        if component_entry.distgit_key == '*' or component_entry.distgit_key == distgit_key:
+            config_dict = merger(component_entry.metadata.primitive(), config_dict)
+
+    return Model(dict_to_model=config_dict)
+
+
+def assembly_basis_event(releases_config: Model, assembly: str) -> typing.Optional[int]:
+    """
+    :param releases_config: The content of releases.yml in Model form.
+    :param assembly: The name of the assembly to assess
+    Returns the basis event for a given assembly. If the assembly has no basis event,
+    None is returned.
+    """
+    if not assembly or not isinstance(releases_config, Model):
+        return None
+
+    target_assembly = releases_config.releases[assembly].assembly
+    if target_assembly.basis.brew_event:
+        return int(target_assembly.basis.brew_event)
+
+    return assembly_basis_event(releases_config, target_assembly.basis.assembly)
```

Comment:
If `target_assembly.basis.assembly` equals to `assembly` or is one of the assemblies this function has visited, this will lead to an infinity recursion.

Would like to prevent some human errors here by changing this function from recursion to iteration and raise an error if there's a cycle that would cause an infinity loop.

Reply:
Added a check.